litestar/contrib/sqlalchemy/repository.py

```diff
@@ -230,26 +232,44 @@ async def get_one_or_none(self, **kwargs: Any) -> ModelT | None:
             The retrieved instance or None
         """
         with wrap_sqlalchemy_exception():
-            statement = kwargs.pop("base_select", self.statement)
+            statement = kwargs.pop("statement", self.statement)
             statement = self._filter_select_by_kwargs(statement=statement, **kwargs)
             instance = (await self._execute(statement)).scalar_one_or_none()
             if instance:
                 self.session.expunge(instance)
             return instance  # type: ignore
 
-    async def get_or_create(self, **kwargs: Any) -> tuple[ModelT, bool]:
+    async def get_or_create(self, match_fields: list[str] | str | None = None, **kwargs: Any) -> tuple[ModelT, bool]:
```

Comment:
After thinking about this, wouldn't this actually just be an upsert that also returns data? Maybe `upsert_returning` or something like that would be a good name.

Reply:
`upsert_returning` definitely isn’t ideal. `get_or_upsert` sounds good,  it actually, it could also just be `upsert` since it’s kinda just an upsert that also returns things, right?

Reply:
We've already got an `upsert` that has a slightly different set of parameters.  Should it just replace that?

Reply:
> wouldn't this actually just be an upsert

I think the kwargs and `match_fields` behavior make it different to the upsert method. Where `upsert()` receives a model instance, and purely matches on identity, this receives and matches on kwargs, and matches on all kwargs by default. Deserves a distinct method IMO and I'm OK with `get_or_create()`.

Reply:
> I think the kwargs and match_fields behavior make it different to the upsert method

Hmm, that's true. But from the way it works, it's still a type of upsert (at least that's what it feels like to me). I do get your point though, and @cofin, I don't think this should replace the already existing `upsert` method. 

How about `match_or_create`, `match_or_upsert`, `match_upsert` or something like that? I just think there ought to be something in the name that indicates this method does not only get or create, but also updates things.

Reply:
The `match` part of this function is completely optional, I should say.  If you omit it, it functions exactly as it used to.  With that in mind, do we still want to rename?

Reply:
> With that in mind, do we still want to rename?

Well.. Personally I still think that it should somehow indicate that it can update things. But I can't really think of a succinct way to phrase that which does not cause confusion with other methods.

In the interest of not blocking this because of a single method name I'd say just leave it as it is :grimacing: